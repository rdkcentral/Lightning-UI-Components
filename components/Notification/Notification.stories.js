import lng from 'wpe-lightning';

import Notification from '.';
import icon from '../../assets/images/ic_lightning_white_32.png';
import mdx from './Notification.mdx';

export default {
  title: 'Notification',
  parameters: {
    docs: {
      page: mdx
    }
  }
};

export const Basic = args =>
  class Basic extends lng.Component {
    static _template() {
      return {
        Notification: {
          type: Notification,
          icon: icon,
          title: 'Notification',
          description: 'This is the description',
          actionArea: {
            text: 'Action Area',
            icon
          }
        }
      };
    }

    _init() {
      setTimeout(() => {
        this.tag('Notification').enter();
      }, 1000);
    }

    _getFocused() {
      return this.tag('Notification');
    }
  };
Basic.args = {
  state: 'enter()'
};
Basic.argTypes = {
  state: {
    control: {
      type: 'radio',
      options: ['enter()', 'dismiss()']
    }
  }
};
Basic.parameters = {
  argActions: {
    state: (state, component) => {
      let func = state.split('(')[0];
      component.tag('Notification')[func]();
    }
  }
};
