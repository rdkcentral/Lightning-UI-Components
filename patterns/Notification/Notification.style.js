export default theme => ({
  minHeight: 120,
  w: 560,
  background: {
    color: theme.palette.background.fill
  },
  backgrounds: theme.palette.background,
  radius: theme.border.radius.small,
  margin: {
    x: theme.spacing(3),
    y: theme.spacing(3)
  },
  icon: {
    size: 64,
    spacing: theme.spacing(2)
  },
  title: {
    ...theme.typography.headline3,
    lineHeight: theme.typography.headline3.lineHeight + 6,
    color: theme.palette.text.light.primary
  },
  description: {
    ...theme.typography.body3,
    lineHeight: theme.typography.body3.lineHeight + 4,
    textColor: theme.palette.text.light.secondary
  },
  actionArea: {
    margin: {
      x: theme.spacing(3),
      y: theme.spacing(2)
    },
    background: {
      color: theme.palette.grey[60],
      h: 64
    },
    text: {
      ...theme.typography.body3,
      lineHeight: theme.typography.body3.lineHeight + 4,
      textColor: theme.palette.text.light.tertiary
    },
    icon: {
      size: 32
    }
  },
  enter: {
    icon: {
      alpha: 1,
      scale: 1
    },
    container: {
      alpha: 1,
      scale: 1
      //x: 0,
      //w: 560
    },
    text: {
      alpha: 1
      //x: 0
    }
  },
  dismiss: {
    icon: {
      alpha: 0.001,
      scale: 0.2
    },
    container: {
      alpha: 0.001,
      scale: 0.4
      //x: 440, // This value is wrong
      //w: 120 // This value is wrong
    },
    text: {
      alpha: 0.001
      //x: 64
    }
  }
});
