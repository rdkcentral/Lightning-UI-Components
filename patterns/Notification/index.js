import lng from '@lightningjs/core';
import MarqueeText from '../../elements/MarqueeText';
import InlineContent from '../../layout/InlineContent';
import Icon from '../../elements/Icon';
import withStyles from '../../mixins/withStyles';
import { withTransitions } from '../../mixins';

export const styles = theme => ({
  minHeight: 120,
  w: 560,
  background: {
    color: theme.palette.background.fill
  },
  backgrounds: theme.palette.background,
  radius: theme.border.radius.small,
  margin: {
    x: theme.spacing(3),
    y: theme.spacing(3)
  },
  shadow: theme.materials.shadow,
  icon: {
    size: 64,
    spacing: theme.spacing(2)
  },
  title: {
    ...theme.typography.headline3,
    lineHeight: theme.typography.headline3.lineHeight + 6,
    color: theme.palette.text.light.primary
  },
  description: {
    ...theme.typography.body3,
    lineHeight: theme.typography.body3.lineHeight + 4,
    textColor: theme.palette.text.light.secondary
  },
  actionArea: {
    margin: {
      x: theme.spacing(3),
      y: theme.spacing(2)
    },
    background: {
      color: theme.palette.grey[60],
      h: 64
    },
    text: {
      ...theme.typography.body3,
      lineHeight: theme.typography.body3.lineHeight + 4,
      textColor: theme.palette.text.light.tertiary
    },
    icon: {
      size: 32
    }
  },
  enter: {
    icon: {
      alpha: 1,
      scale: 1
    },
    container: {
      alpha: 1,
      scale: 1,
      x: 0,
      w: 560
    },
    text: {
      alpha: 1,
      x: 0
    }
  },
  dismiss: {
    icon: {
      alpha: 0.001,
      scale: 0.2
    },
    container: {
      alpha: 0.001,
      scale: 0.4,
      x: 440,
      w: 120
    },
    text: {
      alpha: 0.001,
      x: 64
    }
  }
});

class Notification extends lng.Component {
  static _template() {
    return {
      Container: {
        pivot: 0.5,
        h: this.styles.minHeight,
        w: this.styles.w,
        rect: true,
        clipping: true,
        color: this.styles.background.color,
        shader: {
          type: lng.shaders.RoundedRectangle,
          radius: this.styles.radius
        },
        pivotX: 1,
        Content: {
          x: this.styles.margin.x,
          y: this.styles.margin.y,
          flex: { direction: 'row' },
          Icon: {
            type: Icon,
            w: this.styles.icon.size,
            h: this.styles.icon.size,
            flexItem: { marginRight: this.styles.icon.spacing },
            pivot: 0.5
          },
          Text: {
            w: 432,
            flex: { direction: 'column' },
            Title: {
              type: MarqueeText,
              w: 432,
              h: this.styles.title.lineHeight,
              autoStart: true,
              delay: 3
            },
            Description: {
              type: InlineContent,
              w: 432,
              contentWrap: true,
              textProperties: this.styles.description,
              justify: 'flex-start'
            }
          }
        },
        ActionArea: {
          alpha: 0,
          rect: true,
          color: this.styles.actionArea.background.color,
          h: this.styles.actionArea.background.h,
          w: this.styles.w + 8,
          y: this.styles.minHeight,
          Content: {
            y: this.styles.actionArea.margin.y,
            x: this.styles.actionArea.margin.x,
            w: this.styles.w,
            Text: {
              y: -4,
              type: InlineContent,
              w: 464,
              textProperties: this.styles.actionArea.text,
              justify: 'flex-start'
            },
            Icon: {
              type: Icon,
              x:
                this.styles.w -
                this.styles.actionArea.icon.size -
                this.styles.actionArea.margin.x * 2,
              w: this.styles.actionArea.icon.size,
              h: this.styles.actionArea.icon.size
            }
          }
        }
      }
    };
  }

  _construct() {
    this._whenEnabled = new Promise(
      resolve => (this._enable = resolve),
      console.error
    );
    this._containerH = this.styles.minHeight;
    this.entered = false;
    this._icon = {};
    this._actionArea = false;
    this._contentLoaded = false;
    this._containerHIsFinal = new Promise(resolve => {
      var intervalId = setInterval(() => {
        if (this._contentLoaded) {
          clearInterval(intervalId);
          resolve();
        }
      }, 5);
    });
  }

  _init() {
    this._update();
  }

  set icon(src) {
    if (this._icon.src !== src) {
      this._icon = { ...this.styles.icon, src };
      this._update();
    }
  }

  set title(title) {
    if (this._title !== title) {
      this._title = title;
      this._update();
    }
  }

  set description(description) {
    if (this._description !== description) {
      this._description = description;
      this._update();
    }
  }

  set actionArea(actionArea) {
    if (
      !this._actionArea ||
      (this._actionArea &&
        (this._actionArea.text !== actionArea.text ||
          this._actionArea.icon !== actionArea.icon))
    ) {
      this._actionArea = actionArea;
      this._update();
    }
  }

  $loadedInlineContent() {
    this._contentLoaded = true;
    if (
      this._containerH === this.styles.minHeight &&
      this._Description.multiLineHeight > this.styles.description.lineHeight
    ) {
      this._containerH +=
        this._Description.multiLineHeight - this.styles.description.lineHeight;
      this._update();
    }
  }

  _update() {
    this._whenEnabled.then(() => {
      this._updateProps();
      this._updateIcon();
      this._updateText();
      this._udpateActionArea();
    });
  }

  _updateProps() {
    const state = this.entered ? 'enter' : 'dismiss';

    this._Container.alpha = this.styles[state].container.alpha;
    this._Container.scale = this.styles[state].container.scale;
    this._Container.x = this.styles[state].container.x;
    this._Container.w = this.styles[state].container.w;
  }

  _updateIcon() {
    const state = this.entered ? 'enter' : 'dismiss';
    this._Icon.alpha = this.styles[state].icon.alpha;
    this._Icon.scale = this.styles[state].icon.scale;

    if (this._icon) {
      this._Icon.icon = this._icon.src;
    }
  }

  _updateText() {
    this._Title.title = {
      ...this.styles.title,
      textColor: this.styles.title.color,
      maxLines: 1,
      text: this._title
    };
    this._Description.patch({
      content: this._description
    });

    const descLineLayouter = this._Description.flex._layout._lineLayouter;
    if (
      descLineLayouter &&
      descLineLayouter._lines &&
      descLineLayouter._lines.length
    ) {
      this._containerH =
        this.styles.margin.y * 2 + // top bottom margins
        this.styles.title.lineHeight + // title height
        this._Description.textProperties.lineHeight *
          descLineLayouter._lines.length; // description height
    }

    const state = this.entered ? 'enter' : 'dismiss';
    this._Text.alpha = this.styles[state].text.alpha;
    this._Text.x = this.styles[state].text.x;
  }

  _udpateActionArea() {
    if (this._actionArea && Object.keys(this._actionArea).length) {
      this._ActionArea.alpha = 1;
      this._ActionArea.y = this._containerH;
      this._ActionArea.w = this.styles.w;
      this._ActionArea.tag('Content.Text').content =
        this._actionArea.content || this._actionArea.text;
      this._ActionArea.tag('Content.Icon').icon = this._actionArea.icon;
    } else {
      this._ActionArea.alpha = 0;
    }
  }

  enter() {
    this._containerHIsFinal.then(() => {
      this._udpateActionArea();
      this.entered = true;

      this._Container.h = this._containerH;
      this._Icon.smooth = {
        scale: [this.styles.enter.icon.scale, { duration: 0.18, delay: 0.1 }],
        alpha: [this.styles.enter.icon.alpha, { duration: 0.18, delay: 0.1 }]
      };
      let pop = this._Container.animation({
        duration: 0.34,
        actions: [
          { p: 'alpha', v: { 0: 0, 1: this.styles.enter.container.alpha } },
          {
            p: 'scale',
            v: { 0: 0.4, 0.8: 1.1, 1: this.styles.enter.container.scale }
          }
        ]
      });
      pop.start();
      pop.on('finish', () => {
        this._Container.smooth = {
          w: [this.styles.enter.container.w, { duration: 0.24, delay: 0.2 }],
          x: [this.styles.enter.container.x, { duration: 0.24, delay: 0.2 }]
        };
        this._Text.smooth = {
          x: [this.styles.enter.text.x, { duration: 0.28, delay: 0.3 }],
          alpha: [this.styles.enter.text.alpha, { duration: 0.28, delay: 0.3 }]
        };

        if (this._actionArea && Object.keys(this._actionArea).length) {
          this._Container.smooth = {
            h: [
              this._containerH + this.styles.actionArea.background.h,
              { duration: 0.24, delay: 1 }
            ]
          };
        }
        this.fireAncestors('$notificationEntered');
      });
    });
  }

  dismiss() {
    this.entered = false;
    let delay = 0;
    if (this._actionArea && Object.keys(this._actionArea).length) {
      this._Container.smooth = {
        h: [this._containerH, { duration: 0.18 }]
      };
      delay = 0.18;
    }
    this._Text.smooth = {
      x: [this.styles.dismiss.text.x, { duration: 0.24, delay: delay }],
      alpha: [this.styles.dismiss.text.alpha, { duration: 0.24, delay: delay }]
    };
    this._Container.smooth = {
      w: [
        this.styles.dismiss.container.w,
        { duration: 0.16, delay: delay + 0.1 }
      ],
      x: [
        this.styles.dismiss.container.x,
        { duration: 0.16, delay: delay + 0.1 }
      ],
      scale: [
        this.styles.dismiss.container.scale,
        { duration: 0.16, delay: delay + 0.22 }
      ],
      alpha: [
        this.styles.dismiss.container.alpha,
        { duration: 0.16, delay: delay + 0.28 }
      ]
    };

    this.fireAncestors('$notificationDismissed');
  }

  get _Container() {
    return this.tag('Container');
  }
  get _Icon() {
    return this.tag('Container.Icon');
  }
  get _Text() {
    return this.tag('Container.Text');
  }
  get _Title() {
    return this.tag('Container.Text.Title');
  }
  get _Description() {
    return this.tag('Container.Text.Description');
  }
  get _ActionArea() {
    return this.tag('Container.ActionArea');
  }
}

export default withTransitions(withStyles(Notification, styles));
