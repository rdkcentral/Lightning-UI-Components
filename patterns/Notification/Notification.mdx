import { Canvas, Story } from '@storybook/addon-docs/blocks';
import { Theme } from '../../.storybook/utils';
import Notification from '.';

# Notification

This is an Xfinity branded notification. This notification should only be used inside of Xfinity branded apps. However, this notification can show up as an overlay to external apps if required.

## Usage

`Notification` must take in a `title`, `description`, and `icon`. These are the minimum design requirements for a notification. Notification is also able to accept an optional actionArea object that can provide context on how to proceed with the information provided.

Based on the length of the `description`, the notification will dynamically grow its height to accomodate for any extra lines appended to it. The description field utilizes the [InlineContent](?path=/docs/inlinecontent--basic) component so you are able to mix text and icons as the description.

```js
import { Notification } from '@lightning/ui';

class Example extends lng.Component {
  static _template() {
    return {
      Notification: {
        type: Notification,
        icon: 'path/to/icon',
        title: 'Title',
        description: 'This is the description.'
      }
    };
  }
};
```

<Canvas>
  <Story id="patterns-notification--basic" />
</Canvas>

## API
requirements
### Properties

name|type|required|default|description
--|--|--|--|--
title|string|true|-|Notification title
description|string\|array|true|-|Notification description that can be passed as a string or array depending if icons are needed
icon|string|true|-|URI for icon or inline svg that may live next to the title. See <a href="?path=/docs/icon--basic">Icon</a>
actionArea|object|false|-|object with action area properties
actionArea.text|string\|array|false|-|text for action area can be passed as a string or array depending if icons are needed
actionArea.content|string\|array|false|-|text for action area can be passed as a string or array depending if icons are needed
actionArea.icon|string|false|-|URI for icon in action area

### Methods

#### enter(): null

Triggers enter animation with Xfinity specific curves and duration values. Pops icon in and the slides notification text out to the left. If an action area is present, it will slide out from the bottom of the notification after the initial animation completes.

#### dismiss(): null
Triggers dismiss animation with Xfinity specific curves and duration values. Slides notification back to the right and alpha's all of the elements on the Notification out.

### Styles

Styled components can be adjusted by composing using [withStyles](?path=/docs/mixins-withstyles--basic). Below is the default `styles` object

<Theme theme={Notification.styles} />