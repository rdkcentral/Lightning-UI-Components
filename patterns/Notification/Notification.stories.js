import lng from '@lightningjs/core';

import Notification from '.';
import icon from '../../assets/images/ic_lightning_white_32.png';
import mdx from './Notification.mdx';
import iconSvg from '../../assets/images/success.svg';
export default {
  title: 'Patterns/Notification',
  parameters: {
    docs: {
      page: mdx
    }
  }
};

export const Basic = args =>
  class Basic extends lng.Component {
    static _template() {
      return {
        Notification: {
          type: Notification,
          icon: iconSvg,
          title:
            'Notification: Lorem ipsum dolor sit amet, consectetur adipiscing elit',
          description:
            'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt',
          actionArea: {
            text:
              'Action Area: Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt',
            icon
          },
          onEnter() {
            this.dismiss();
          }
        }
      };
    }

    _init() {
      setTimeout(() => {
        this.tag('Notification').enter();
      }, 1e3);
    }

    $notificationEntered() {
      console.log('Notification entered');
    }

    $notificationDismissed() {
      console.log('Notification dismissed');
    }

    _getFocused() {
      return this.tag('Notification');
    }
  };
Basic.args = {
  state: 'enter()',
  actionArea: {
    text:
      'Action Area: Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt',
    icon
  }
};
Basic.argTypes = {
  state: {
    control: {
      type: 'radio',
      options: ['enter()', 'dismiss()']
    }
  },
  actionArea: { control: 'boolean' }
};
Basic.parameters = {
  argActions: {
    state: (state, component) => {
      let func = state.split('(')[0];
      component.tag('Notification')[func]();
    },
    actionArea: (action, component) => {
      component.tag('Notification').actionArea = action
        ? {
            text:
              'Action Area: Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt',
            icon
          }
        : false;
    }
  }
};
