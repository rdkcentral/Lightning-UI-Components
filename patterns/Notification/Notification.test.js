import TestUtils from '../../test/lightning-test-utils';
import Notification from '.';

const createNotification = TestUtils.makeCreateComponent(Notification);

describe('Notification', () => {
  let notification, testRenderer;

  beforeEach(() => {
    [notification, testRenderer] = createNotification({
      title: 'Title',
      description: 'This is a description',
      icon: 'patch/to/src'
    });
  });
  afterEach(() => {
    notification = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('sets an icon', () => {
    expect(notification._icon).toEqual('patch/to/src');
  });

  it('will not show the action area if it is undefined', () => {
    [notification, testRenderer] = createNotification({
      title: 'Title',
      description: 'This is a description',
      icon: 'patch/to/src'
    });
    expect(notification._ActionArea.alpha).toBe(0.001);
  });

  it('will show the action area if one is passed in', done => {
    [notification, testRenderer] = createNotification({
      title: 'Title',
      description: 'This is a description',
      icon: 'patch/to/src',
      actionArea: {
        text: 'Action Area'
      }
    });
    expect(notification._actionArea).toEqual({
      text: 'Action Area'
    });

    notification.enter();

    testRenderer.update();

    setTimeout(() => {
      expect(notification._ActionArea.alpha).toBe(1);
      done();
    }, 1000); // bumping the timer resolves a race condition with the alpha update
  });

  it('should not show the action area if the one passed in does not have text', async () => {
    [notification, testRenderer] = createNotification({
      title: 'Title',
      description: 'This is a description',
      icon: 'patch/to/src',
      actionArea: {}
    });

    await notification.enter();

    expect(notification._ActionArea.alpha).toEqual(0.001);
  });

  it('should be able to mount entered', done => {
    [notification, testRenderer] = createNotification({
      title: 'Title',
      description: 'This is a description',
      icon: 'patch/to/src',
      entered: true
    });
    setTimeout(() => {
      expect(notification._Container.alpha).toEqual(
        notification.styles.enter.container.alpha
      );
      expect(notification._Container.scale).toEqual(
        notification.styles.enter.container.scale
      );
      expect(notification._Icon.alpha).toEqual(
        notification.styles.enter.icon.alpha
      );
      expect(notification._Icon.scale).toEqual(
        notification.styles.enter.icon.scale
      );
      expect(notification._Text.alpha).toEqual(
        notification.styles.enter.text.alpha
      );
      done();
    }, 1500);
  });

  it('should call enter() and complete transitions', done => {
    notification.enter();
    setTimeout(() => {
      expect(notification._Container.alpha).toEqual(
        notification.styles.enter.container.alpha
      );
      expect(notification._Container.scale).toEqual(
        notification.styles.enter.container.scale
      );
      expect(notification._Icon.alpha).toEqual(
        notification.styles.enter.icon.alpha
      );
      expect(notification._Icon.scale).toEqual(
        notification.styles.enter.icon.scale
      );
      expect(notification._Text.alpha).toEqual(
        notification.styles.enter.text.alpha
      );
      done();
    }, 1500);
  });

  it('should call dismiss() and complete transitions', done => {
    notification.dismiss();
    setTimeout(() => {
      expect(notification._Container.alpha).toEqual(
        notification.styles.dismiss.container.alpha
      );
      expect(notification._Container.scale).toEqual(
        notification.styles.dismiss.container.scale
      );
      expect(notification._Icon.alpha).toEqual(
        notification.styles.dismiss.icon.alpha
      );
      expect(notification._Text.alpha).toEqual(
        notification.styles.dismiss.text.alpha
      );
      done();
    }, 2000);
  });
});
