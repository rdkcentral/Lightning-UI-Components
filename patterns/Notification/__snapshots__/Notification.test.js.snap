// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification renders 1`] = `
Object {
  "Component": Object {
    "Container": Object {
      "active": false,
      "alpha": 0.001,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "ActionArea": Object {
          "active": false,
          "alpha": 0.001,
          "attached": true,
          "boundsMargin": null,
          "children": Object {
            "Content": Object {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "children": Object {
                "Text": Object {
                  "active": false,
                  "alpha": 1,
                  "attached": true,
                  "boundsMargin": null,
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": false,
                  "flexItem": false,
                  "h": 0,
                  "hasFinalFocus": false,
                  "hasFocus": false,
                  "isComponent": true,
                  "mount": 0,
                  "mountX": 0,
                  "mountY": 0,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Text",
                  "renderOfScreen": undefined,
                  "renderToTexture": false,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": "",
                  "tag": [Function],
                  "tags": Array [
                    "Text",
                  ],
                  "type": "InlineContent",
                  "visible": true,
                  "w": 512,
                  "x": 0,
                  "y": -4,
                  "zIndex": 0,
                },
              },
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Content",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": Array [
                "Content",
              ],
              "visible": true,
              "w": 512,
              "x": 24,
              "y": 16,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": "ff3e3e45",
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "ActionArea",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "ActionArea",
          ],
          "texture": Object {
            "type": "RectangleTexture",
          },
          "visible": true,
          "w": 560,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Content": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": Object {
            "Icon": Object {
              "active": false,
              "alpha": 0.001,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": true,
              "h": 64,
              "hasFinalFocus": false,
              "hasFocus": false,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Icon",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "tags": Array [
                "Icon",
              ],
              "texture": Object {
                "src": "patch/to/src",
                "type": "ImageTexture",
              },
              "type": "Icon",
              "visible": true,
              "w": 64,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
            "Text": Object {
              "active": false,
              "alpha": 0.001,
              "attached": true,
              "boundsMargin": null,
              "children": Object {
                "Description": Object {
                  "active": false,
                  "alpha": 1,
                  "attached": true,
                  "boundsMargin": null,
                  "children": Object {
                    "Element-12": Object {
                      "active": false,
                      "alpha": 1,
                      "attached": true,
                      "boundsMargin": null,
                      "clipping": false,
                      "color": 4294967295,
                      "enabled": true,
                      "flex": false,
                      "flexItem": true,
                      "h": 40,
                      "isComponent": undefined,
                      "mount": 0,
                      "mountX": 0,
                      "mountY": 0,
                      "pivot": 0.5,
                      "pivotX": 0.5,
                      "pivotY": 0.5,
                      "ref": null,
                      "renderOfScreen": undefined,
                      "renderToTexture": false,
                      "scale": 1,
                      "scaleX": 1,
                      "scaleY": 1,
                      "state": undefined,
                      "tag": [Function],
                      "texture": Object {
                        "fontFace": "XfinityStandardMedium",
                        "fontSize": 24,
                        "h": 40,
                        "lineHeight": 36,
                        "text": "This ",
                        "textColor": 3439329279,
                        "type": "TextTexture",
                        "verticalAlign": "middle",
                      },
                      "visible": true,
                      "w": 0,
                      "x": 0,
                      "y": 0,
                      "zIndex": 0,
                    },
                    "Element-13": Object {
                      "active": false,
                      "alpha": 1,
                      "attached": true,
                      "boundsMargin": null,
                      "clipping": false,
                      "color": 4294967295,
                      "enabled": true,
                      "flex": false,
                      "flexItem": true,
                      "h": 40,
                      "isComponent": undefined,
                      "mount": 0,
                      "mountX": 0,
                      "mountY": 0,
                      "pivot": 0.5,
                      "pivotX": 0.5,
                      "pivotY": 0.5,
                      "ref": null,
                      "renderOfScreen": undefined,
                      "renderToTexture": false,
                      "scale": 1,
                      "scaleX": 1,
                      "scaleY": 1,
                      "state": undefined,
                      "tag": [Function],
                      "texture": Object {
                        "fontFace": "XfinityStandardMedium",
                        "fontSize": 24,
                        "h": 40,
                        "lineHeight": 36,
                        "text": "is ",
                        "textColor": 3439329279,
                        "type": "TextTexture",
                        "verticalAlign": "middle",
                      },
                      "visible": true,
                      "w": 0,
                      "x": 0,
                      "y": 0,
                      "zIndex": 0,
                    },
                    "Element-14": Object {
                      "active": false,
                      "alpha": 1,
                      "attached": true,
                      "boundsMargin": null,
                      "clipping": false,
                      "color": 4294967295,
                      "enabled": true,
                      "flex": false,
                      "flexItem": true,
                      "h": 40,
                      "isComponent": undefined,
                      "mount": 0,
                      "mountX": 0,
                      "mountY": 0,
                      "pivot": 0.5,
                      "pivotX": 0.5,
                      "pivotY": 0.5,
                      "ref": null,
                      "renderOfScreen": undefined,
                      "renderToTexture": false,
                      "scale": 1,
                      "scaleX": 1,
                      "scaleY": 1,
                      "state": undefined,
                      "tag": [Function],
                      "texture": Object {
                        "fontFace": "XfinityStandardMedium",
                        "fontSize": 24,
                        "h": 40,
                        "lineHeight": 36,
                        "text": "a ",
                        "textColor": 3439329279,
                        "type": "TextTexture",
                        "verticalAlign": "middle",
                      },
                      "visible": true,
                      "w": 0,
                      "x": 0,
                      "y": 0,
                      "zIndex": 0,
                    },
                    "Element-15": Object {
                      "active": false,
                      "alpha": 1,
                      "attached": true,
                      "boundsMargin": null,
                      "clipping": false,
                      "color": 4294967295,
                      "enabled": true,
                      "flex": false,
                      "flexItem": true,
                      "h": 40,
                      "isComponent": undefined,
                      "mount": 0,
                      "mountX": 0,
                      "mountY": 0,
                      "pivot": 0.5,
                      "pivotX": 0.5,
                      "pivotY": 0.5,
                      "ref": null,
                      "renderOfScreen": undefined,
                      "renderToTexture": false,
                      "scale": 1,
                      "scaleX": 1,
                      "scaleY": 1,
                      "state": undefined,
                      "tag": [Function],
                      "texture": Object {
                        "fontFace": "XfinityStandardMedium",
                        "fontSize": 24,
                        "h": 40,
                        "lineHeight": 36,
                        "text": "description",
                        "textColor": 3439329279,
                        "type": "TextTexture",
                        "verticalAlign": "middle",
                      },
                      "visible": true,
                      "w": 0,
                      "x": 0,
                      "y": 0,
                      "zIndex": 0,
                    },
                  },
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": true,
                  "flexItem": true,
                  "h": 0,
                  "hasFinalFocus": false,
                  "hasFocus": false,
                  "isComponent": true,
                  "mount": 0,
                  "mountX": 0,
                  "mountY": 0,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Description",
                  "renderOfScreen": undefined,
                  "renderToTexture": false,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": "",
                  "tag": [Function],
                  "tags": Array [
                    "Description",
                  ],
                  "type": "InlineContent",
                  "visible": true,
                  "w": 424,
                  "x": 0,
                  "y": 0,
                  "zIndex": 0,
                },
                "Title": Object {
                  "active": false,
                  "alpha": 0.001,
                  "attached": true,
                  "boundsMargin": null,
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": false,
                  "flexItem": true,
                  "h": 0,
                  "hasFinalFocus": false,
                  "hasFocus": false,
                  "isComponent": true,
                  "mount": 0,
                  "mountX": 0,
                  "mountY": 0,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Title",
                  "renderOfScreen": undefined,
                  "renderToTexture": false,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": "",
                  "tag": [Function],
                  "tags": Array [
                    "Title",
                  ],
                  "texture": Object {
                    "fontFace": "XfinityBrownBold",
                    "fontSize": 28,
                    "letterSpacing": -0.2,
                    "lineHeight": 36,
                    "text": "Title",
                    "type": "TextTexture",
                    "verticalAlign": "middle",
                    "wordWrapWidth": 424,
                  },
                  "type": "TextBox",
                  "visible": true,
                  "w": 0,
                  "x": 0,
                  "y": 0,
                  "zIndex": 0,
                },
              },
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": true,
              "flexItem": true,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Text",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": Array [
                "Text",
              ],
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Content",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Content",
          ],
          "visible": true,
          "w": 0,
          "x": 24,
          "y": 24,
          "zIndex": 0,
        },
      },
      "clipping": true,
      "color": "ff232328",
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 112,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Container",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 0.4,
      "scaleX": 0.4,
      "scaleY": 0.4,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Container",
      ],
      "texture": Object {
        "type": "RectangleTexture",
      },
      "visible": true,
      "w": 112,
      "x": 448,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
